src/gordon_gcp/plugins/publisher.py

```diff
@@ -13,44 +13,222 @@
             :interface:`interfaces.IEventMessages` that were not
             processed due to problems.
+        http_client (http.AIOConnection):
+            A Google HTTP connection class.
     """
     phase = 'publish'
 
-    def __init__(self, config, success_channel, error_channel):
+    def __init__(self, config, success_channel, error_channel, http_client):
         self.config = config
         self.success_channel = success_channel
         self.error_channel = error_channel
+        self.http_client = http_client
+        self.publish_wait_timeout = config['publish_wait_timeout']
+        self.managed_zone = config['managed_zone']
+        self.dns_zone = config['dns_zone']
+        self.project = config['project']
+        self.default_ttl = self.config['default_ttl']
+        api_version = config.get('version', 'v1')
+        self.base_changes_url = BASE_CHANGES_ENDPOINT.format(
+            host=HOST, version=api_version, project=self.project,
+            managedZone=self.managed_zone)
+        self.resource_records_url = RESOURCE_RECORDS_ENDPOINT.format(
+            host=HOST, project=self.project, managedZone=self.managed_zone)
+        self._logger = logging.getLogger('')
+
+    # TODO: This will be eventually moved to GEventMessage
+    async def update_phase(self, event_msg, phase=None):
+        old_phase = event_msg.phase
+        event_msg.phase = phase or self.phase
+        msg = f'Updated phase from "{old_phase}" to "{event_msg.phase}".'
+        event_msg.append_to_history(msg, self.phase)
+
+    def _format_changes(self, action, resource_record):
+        """Return dict containing the changes to be made.
+
+        Args:
+            action (str): The action of the changes to publish;
+                can be one of ['additions', 'deletions'].
+            resource_record (dict): Dict containing record data.
+
+        Returns:
+            The changes (dict) to post to the Google API.
+        """
+        resource_record['kind'] = 'dns#resourceRecordSet'
+        resource_record['ttl'] = resource_record.get('ttl', self.default_ttl)
+        return {
+            'kind': 'dns#change',
+            action: [resource_record]
+        }
+
+    async def _publish_changes(self, changes):
+        """Publish changes to the Google DNS Changes API.
+
+        Args:
+            changes (dict): The DNS record changes to publish.
+
+        Returns:
+            change_id (str): Change ID of the Changes resource.
+
+        Raises:
+            GCPHTTPConflictError on HTTP 409.
+        """
+        try:
+            resp = await self.http_client.request(
+                'post', self.base_changes_url, json=changes)
+        except exceptions.GCPHTTPError as e:
+            error_msg = e.args[0]
+            # Extract status code from error message.
+            match = re.search('\d+', error_msg)
+            if match:
+                status_code = int(match.group(0))
+                if status_code == 409:
```

Comment:
I dunno if you already talked about this while mobbing, but this is a future-refactoring idea: it seems awkward that the publisher has to parse an error string to figure out why the http client is failing. Since the publisher also has to know a bit about HTTP to know when to drop/retry, maybe the http client could add the status code to the exception?

Reply:
We did talk about it a bit, but we didn't come up with this particular idea, which I think is a good one. Mostly we talked about why the http client does a `raise_for_status` at all, and whether just returning the response (which has the code) would be better (or at least having it configurable).  I think the decision we settled on was that some refactoring was a good idea,  but outside the scope of getting this PR merged.